### PR TITLE
Abillity to disable passes in the frontend and midend (#830)

### PR DESCRIPTION
**Pull request by marko-stanoj**
Added four new compiler options :
--listFrontendPasses, prints names of all frontend passes
--listMidendPasses, prints names of all midend passes
--excludeFrontendPasses,  exclude one or more frontend passes,
			  which names are passed as arguments, from further processing
--excludeMidendPasses, exclude one or more midend passes,
                       which names are passed as arguments, from further processing

*Developers should be aware of possibility that skipping some passes may trigger bugs
 in other parts of compiler

### CODE REVIEW

File: backends/bmv2/psa_switch/midend.cpp [hunk 1]
```diff
@@ -137,7 +139,19 @@ PsaSwitchMidEnd::PsaSwitchMidEnd(CompilerOptions& options) : MidEnd(options) {
...
             new VisitFunctor([this, evaluator]() { toplevel = evaluator->getToplevelBlock(); }),
-        });
+        };
+        if (options.listMidendPasses) {
+            for (auto it : midendPasses) {
+                if (it != nullptr) {
+                *outStream << it->name() <<'\n';
+                }
+            }
+        return;
```
Comment (mbudiu-vmw) [inline backends/bmv2/psa_switch/midend.cpp]: indentation

File: backends/bmv2/simple_switch/options.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ public:
+    SimpleSwitchOptions() {
+        registerOption("--listMidendPasses", nullptr,
+                [this](const char*) {
+                    listMidendPasses = true;
+                    loadIRFromJson = false;
+                    SimpleSwitchMidEnd midEnd(*this, outStream);
+                    exit(0);
+                    return false; },
+                "[SimpleSwtitch back-end] Lists exact name of all midend passes.\n");
```
Comment (mbudiu-vmw) [inline backends/bmv2/simple_switch/options.h]: typo

File: backends/bmv2/psa_switch/options.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+namespace BMV2 {
+
+class PsaSwitchOptions : public BMV2Options {
+ public:
+    PsaSwitchOptions() {
+        registerOption("--listMidendPasses", nullptr,
+                [this](const char*) {
+                    listMidendPasses = true;
+                    loadIRFromJson = false;
+                    PsaSwitchMidEnd midEnd(*this, outStream);
```
Comment (mbudiu-vmw) [inline backends/bmv2/psa_switch/options.h]: this is a bit strange; I would have expected that you will pass the flag to the existing midend.
But I guess that this works.

File: backends/ebpf/midend.cpp [hunk 4]
```diff
@@ -104,7 +105,19 @@ const IR::ToplevelBlock* MidEnd::run(EbpfOptions& options, const IR::P4Program*
             new EBPF::Lower(&refMap, &typeMap),
             evaluator,
             new P4::MidEndLast()
-       };
+        };
+        if (options.listMidendPasses) {
+            for (auto it : midendPasses) {
+                if (it != nullptr) {
+                *outStream << it->name() <<'\n';
```
Comment (mbudiu-vmw) [inline backends/ebpf/midend.cpp]: indentation

File: backends/bmv2/psa_switch/midend.h [hunk 5]
```diff
@@ -27,7 +27,7 @@ namespace BMV2 {
 
 class PsaSwitchMidEnd : public MidEnd {
  public:
-    explicit PsaSwitchMidEnd(CompilerOptions& options);
+    explicit PsaSwitchMidEnd(CompilerOptions& options, std::ostream* outStream = nullptr);
```
Comment (mbudiu-vmw) [inline backends/bmv2/psa_switch/midend.h]: you should probably add a comment explaining what the outStream is for.

File: backends/ebpf/ebpfOptions.cpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+                    loadIRFromJson = false;
+                    listMidendPasses = true;
+                    EBPF::MidEnd midend;
+                    midend.run(*this, nullptr, outStream);
+                    exit(0);
+                    return false; },
+                "[ebpf back-end] Lists exact name of all midend passes.\n");
+        registerOption("--fromJSON", "file",
+                [this](const char* arg) { loadIRFromJson = true; file = arg; return true; },
+                "Use IR representation from JsonFile dumped previously,"\
```
Comment (mbudiu-vmw) [inline backends/ebpf/ebpfOptions.cpp]: I don't think you need the backslash here

File: backends/ebpf/midend.h [hunk 7]
```diff
@@ -31,7 +31,8 @@ class MidEnd {
     P4::TypeMap            typeMap;
 
     void addDebugHook(DebugHook hook) { hooks.push_back(hook); }
-    const IR::ToplevelBlock* run(EbpfOptions& options, const IR::P4Program* program);
+    const IR::ToplevelBlock* run(EbpfOptions& options,
+    const IR::P4Program* program, std::ostream* outStream = nullptr);
```
Comment (mbudiu-vmw) [inline backends/ebpf/midend.h]: indentation is wrong

File: frontends/p4/frontend.cpp [hunk 8]
```diff
@@ -193,10 +192,25 @@ const IR::P4Program *FrontEnd::run(const CompilerOptions &options, const IR::P4P
         new HierarchicalNames(),
         new FrontEndLast(),
     };
+    if (options.listFrontendPasses) {
+        for (auto it : frontendPasses) {
+            if (it != nullptr) {
+            *outStream << it->name() <<'\n';
```
Comment (mbudiu-vmw) [inline frontends/p4/frontend.cpp]: indentation

**Comment by mbudiu-vmw** [comment]
Fixes #830

**Comment by marko-stanoj** [comment]
Hello @mbudiu-vmw ,
Thank you for your time and comments.
I updated my PR and I'm looking forward to your response.

File: backends/bmv2/psa_switch/options.h (hunk 3 above)
Comment (marko-stanoj) [reply backends/bmv2/psa_switch/options.h]: I assumed that p4c should be able to run with single option '--listMidendPasses', similar to '--help' option. 
If so, there is no input file to compile and the best way to get to the midend passes names 
instantly is to call midend constructor.

**Pull request merged**